src/mono/mono/eventpipe/ep-rt-mono.c

```diff
@@ -2857,6 +2857,433 @@ ep_rt_mono_write_event_ee_startup_start (void)
+		(s_name_len + 1) * sizeof(char) + 			// Size of name, including null terminator
+		bulk_type_value->c_type_parameters * sizeof(uint64_t);	// Type parameters
+}
+
+// ETW has a limitation of 64K for TOTAL event Size, however there is overhead associated with
+// the event headers.   It is unclear exactly how much that is, but 1K should be sufficiently
+// far away to avoid problems without sacrificing the perf of bulk processing.
+static const uint32_t cb_max_etw_event = 63 * 1024;
+
+// Estimate of how many bytes we can squeeze in the event data for the value struct
+// array.  (Intentionally overestimate the size of the non-array parts to keep it safe.)
+static const uint32_t k_max_bytes_type_values = (cb_max_etw_event - 0x30);
+
+// Estimate of how many type value elements we can put into the struct array, while
+// staying under the ETW event size limit. Note that this is impossible to calculate
+// perfectly, since each element of the struct array has variable size.
+//
+// In addition to the byte-size limit per event, Windows always forces on us a
+// max-number-of-descriptors per event, which in the case of BulkType, will kick in
+// far sooner. There's a max number of 128 descriptors allowed per event. 2 are used
+// for Count + ClrInstanceID. Then 4 per batched value. (Might actually be 3 if there
+// are no type parameters to log, but let's overestimate at 4 per value).
+static const uint32_t k_max_count_type_values = (128 - 2) / 4;
+
+struct _BulkTypeEventLogger {
+	BulkTypeValue m_rg_bulk_type_values[k_max_count_type_values];
+	uint32_t m_n_bulk_type_value_count;
+	uint32_t m_n_bulk_type_value_byte_count;
+	uint8_t *m_p_bulk_type_event_buffer[65536];
+};
+
+//---------------------------------------------------------------------------------------
+//
+// ep_rt_mono_fire_bulk_type_event fires an ETW event for all the types batched so far,
+// it then resets the state to start batching new types at the beginning of the
+// m_rg_bulk_type_values array.
+//
+
+void
+ep_rt_mono_fire_bulk_type_event (BulkTypeEventLogger *p_type_logger)
+{
+	if (p_type_logger->m_n_bulk_type_value_count == 0)
+		return;
+	memset(p_type_logger->m_p_bulk_type_event_buffer, 0, 65536 * sizeof(uint8_t));
+	uint16_t n_clr_instance_id = clr_instance_get_id();
+
+	uint32_t i_size = 0;
+
+	for (int i_type_data = 0; i_type_data < p_type_logger->m_n_bulk_type_value_count; i_type_data++)
+	{
+		BulkTypeValue *target = &p_type_logger->m_rg_bulk_type_values[i_type_data];
+
+		memcpy(p_type_logger->m_p_bulk_type_event_buffer + i_size,
+			   &target->fixed_sized_data,
+			   sizeof(target->fixed_sized_data));
+		i_size += sizeof(target->fixed_sized_data);
+
+		char *wsz_name = target->s_name;
+		if (!wsz_name || strlen(wsz_name) == 0)
+		{
+			p_type_logger->m_p_bulk_type_event_buffer[i_size++] = 0;
+			p_type_logger->m_p_bulk_type_event_buffer[i_size++] = 0;
+		}
+		else
+		{
+			uint32_t name_size = (strlen(target->s_name) + 1) * sizeof(wchar_t);
```

Comment:
Thanks for the catch :). I haven't gotten around to trying to implement https://github.com/dotnet/runtime/blob/ac6e77b676068c13878aa1a89bdccfaabde50d35/src/coreclr/vm/eventtrace.cpp#L1880-L1896. As far as I can tell, thats the only instance where s_name would have been set, so this `else` actually isn't hit at all until s_name is set to something. 

I'll change the calculation to be `sizeof(wchar_t)`.

Note: I was somehow getting a `target->s_name` equivalent to `""` instead of it staying as `NULL`. so there are some additions to address that, such as 
```
	uint32_t s_name_len = 0;
	if (bulk_type_value->s_name)
		s_name_len = strlen(bulk_type_value->s_name);
```
and checking the length `if (!wsz_name || strlen(wsz_name) == 0)`

Reply:
Changed the calculation of `ep_rt_mono_get_byte_count_in_event` to use `wchar_t`, also I realized that `s_name`, though never being set yet because I haven't implemented the snippet above, may contain junk from the g_malloc. Addressed that in the BulkTypeValue init `ep_rt_bulk_type_value_init`